packages/graphql-codegen-cli/src/codegen.ts

```diff
@@ -65,87 +68,211 @@ async function mergeSchemas(schemas: GraphQLSchema[]): Promise<GraphQLSchema> {
+    }
   }
 
-  /* Load root schemas */
-  const rootSchema = await mergeSchemas(
-    await Promise.all(schemas.map(pointToScehma => loadSchema(pointToScehma, config)))
-  );
-
-  /* Load root documents */
-  let rootDocuments: DocumentFile[] = [];
+  async function loadRootSchema() {
+    /* Load root schemas */
+    rootSchema = await mergeSchemas(await Promise.all(schemas.map(pointToScehma => loadSchema(pointToScehma, config))));
+  }
 
-  if (documents.length > 0) {
-    const foundDocumentsPaths = await documentsFromGlobs(documents);
-    rootDocuments = await loadDocumentsSources(foundDocumentsPaths);
+  async function loadRootDocuments() {
+    /* Load root documents */
+    if (documents.length > 0) {
+      const foundDocumentsPaths = await documentsFromGlobs(documents);
+      rootDocuments = await loadDocumentsSources(foundDocumentsPaths);
 
-    if (rootSchema) {
-      const errors = validateGraphQlDocuments(rootSchema, rootDocuments);
-      checkValidationErrors(errors, !config.watch);
+      if (rootSchema) {
+        const errors = validateGraphQlDocuments(rootSchema, rootDocuments);
+        checkValidationErrors(errors);
+      }
     }
   }
 
-  /* Iterate through all output files, and execute each template piece */
-  for (let filename of Object.keys(generates)) {
-    const outputConfig = generates[filename];
-    const outputFileTemplateConfig = outputConfig.config || {};
-    let outputSchema = rootSchema;
-    let outputDocuments: DocumentFile[] = rootDocuments;
-
-    const outputSpecificSchemas = normalizeInstanceOrArray<Types.Schema>(outputConfig.schema);
-    if (outputSpecificSchemas.length > 0) {
-      outputSchema = await mergeSchemas([
-        rootSchema,
-        ...(await Promise.all(outputSpecificSchemas.map(pointToScehma => loadSchema(pointToScehma, config))))
-      ]);
-    }
-
-    const outputSpecificDocuments = normalizeInstanceOrArray<Types.OperationDocument>(outputConfig.documents);
-
-    if (outputSpecificDocuments.length > 0) {
-      const foundDocumentsPaths = await documentsFromGlobs(outputSpecificDocuments);
-      const additionalDocs = await loadDocumentsSources(foundDocumentsPaths);
+  listr.add({
+    title: 'Parse configuration',
+    task: ctx => {
+      normalize();
 
-      if (outputSchema) {
-        const errors = validateGraphQlDocuments(outputSchema, additionalDocs);
-        checkValidationErrors(errors, !config.watch);
+      ctx.hasSchemas = schemas.length > 0;
+      ctx.hasDocuments = documents.length > 0;
+    }
+  });
+
+  listr.add({
+    title: 'Load schema',
+    enabled: ctx => ctx.hasSchemas,
+    task: loadRootSchema
+  });
+
+  listr.add({
+    title: 'Load documents',
+    enabled: ctx => ctx.hasDocuments,
+    task: async () => {
+      try {
+        await loadRootDocuments();
+      } catch (error) {
+        if (!config.watch) {
+          throw error;
+        } else {
+          cliError(error, true);
```

Comment:
I think we should have here something different in watch mode, not process.exit but some info